### DISCUSSION TITLE
Documentation, Erasure Encoding Question & Support

### DISCUSSION THREAD

**Discussion by BloodBlight** [seaweedfs/seaweedfs#2552] [Q&A]
I am just starting to poke at SeaweedFS as a replacement for LizardsFS (A MooseFS 2.0 fork), and possibly MooseFS V3.  So I don’t know a lot about seaweed just yet, but am looking at it for several possible projects.

In the Architecture section of:
https://github.com/chrislusf/seaweedfs/wiki/Erasure-coding-for-warm-storage

"SeaweedFS implemented 10.4 Reed-Solomon Erasure Coding (EC). The large volumes are split into chunks of 1GB, and every 10 data chunks are also encoded into 4 parity chunks. So a 30 GB data volume will be encoded into 14 EC shards, each shard is of size 3 GB and has 3 EC blocks.”

If large volumes are split into 1GB chunks, why would they be encoded into 14 3GB chunks?  Wouldn’t this be encoded into 42 1GB chunks?

Later it mentions that only volumes over 95% full will be encoded.  Is this the entire volume, or the 1GB chunks?  Is SeaweedFS a COW file system (maybe some sort of hybrid), does it (or will it) support snapshots?  What if one of the files in these these encoded volumes is updated or deleted.  More information on this process would be greatly appreciated.

In the side note section it it mentioned that the encoding can be adjusted.  It looks like this is in the source code and would have to be compiled to adjust it.  What happens if this is changed, and then the binaries are updated but have not been adjusted.  For example, someone custom compiles a version with an EC of 7D and 3P, but then installs the stock version.  Will it still be able to read the data, will re-sealing the data allow already encoded data to be adjusted?  Will there ever be an option to have multiple policies?  I guess that could be done with multiple instances...

It also mentions supporting SeaweedFS first.  Are you referring to Patreon, or will you be offering direct support?  Home lab level, Enterprise?

Sorry for the wall of questions, but the erasure encoding is an important feature to me.

**Comment by chrislusf** [comment]
> If large volumes are split into 1GB chunks, why would they be encoded into 14 3GB chunks? Wouldn’t this be encoded into 42 1GB chunks?

The 3 chunks are merged into one file.

> Later it mentions that only volumes over 95% full will be encoded. Is this the entire volume, or the 1GB chunks?

The entire volume. The 95% is something you can control via admin script.



> Is SeaweedFS a COW file system (maybe some sort of hybrid), does it (or will it) support snapshots?

Not supporting snapshots yet.

> What if one of the files in these these encoded volumes is updated or deleted.

COW. The updated or deleted entries are marked as deleted, and will be purged during compaction.



> In the side note section it it mentioned that the encoding can be adjusted. It looks like this is in the source code and would have to be compiled to adjust it.

Yes. It is not meant to be adjusted freely.

> It also mentions supporting SeaweedFS first. Are you referring to Patreon, or will you be offering direct support? Home lab level, Enterprise?

Usually customized EC is for enterprise.

**Reply by BloodBlight** [reply]
> Usually customized EC is for enterprise.

I was hoping you could go into more detail as to what the means.  Are you implying a local development team, or as a service provided by contract.  Or both.

**Reply by chrislusf** [reply]
This would need a contract to fund the project.
